engine/src/main/scala/cromwell/engine/workflow/JobInputEvaluator.scala

```diff
@@ -0,0 +1,176 @@
+  private def unqualifiedInputsFromInputFile(call: Call): Map[LocallyQualifiedName, WdlValue] = splitInputs collect {
+    case((root, inputName), v) if root == call.fullyQualifiedName => inputName -> v
+  }
+
+  // In this case, the lookup function is effectively equivalent to looking into unqualifiedWorkflowInputs for the value
+  // because the resolution / evaluation / coercion has already happened in the MaterializeWorkflowDescriptorActor
+  private def lookupWorkflowDeclaration(identifier: String) = {
+    unqualifiedWorkflowInputs.get(identifier) match {
+      case Some(value) => Success(value)
+      case None => Failure(new WdlExpressionException(s"Could not resolve variable $identifier as a workflow input"))
+    }
+  }
+
+  private def lookupCall(key: BackendJobDescriptorKey, outputStore: OutputStore)(name: String): Try[WdlCallOutputsObject] = {
+    calls find { _.unqualifiedName == name } match {
+      case Some(matchedCall) =>
+        /**
+          * After matching the Call, this determines if the `key` depends on a single shard
+          * of a scatter'd job or if it depends on the whole thing.  Right now, the heuristic
+          * is "If we're both in a scatter block together, then I depend on a shard.  If not,
+          * I depend on the collected value"
+          *
+          * TODO: nested-scatter - this will likely not be sufficient for nested scatters
+          */
+        val index: ExecutionIndex = matchedCall.closestCommonAncestor(key.scope) flatMap {
+          case s: Scatter => key.index
+          case _ => None
+        }
+        fetchCallOutputEntries(matchedCall, index, outputStore)
+      case None => Failure(new WdlExpressionException(s"Could not find a call with name '$name'"))
+    }
+  }
+
+  private def fetchCallOutputEntries(call: Call, index: ExecutionIndex, outputStore: OutputStore): Try[WdlCallOutputsObject] = {
+    def outputEntriesToMap(outputs: Traversable[OutputEntry]): Map[String, Try[WdlValue]] = {
+      outputs map { output =>
+        output.wdlValue match {
+          case Some(wdlValue) => output.name -> Success(wdlValue)
+          case None => output.name -> Failure(new RuntimeException(s"Could not retrieve output ${output.name} value"))
+        }
+      } toMap
+    }
+
+    outputStore.get(OutputCallKey(call, index)) match {
+      case Some(outputs) =>
+        TryUtil.sequenceMap(outputEntriesToMap(outputs), s"Output fetching for call ${call.unqualifiedName}") map { outputsMap =>
+          WdlCallOutputsObject(call, outputsMap)
+        }
+      case None => Failure(new RuntimeException(s"Could not find call ${call.unqualifiedName}"))
+    }
+  }
+
+  private def buildInputMappingsLookup(jobKey: BackendJobDescriptorKey, outputStore: OutputStore)(identifier: String): WdlValue = {
+    // Will need to be enhanced with scatter variable resolvers
+    val resolvers = Stream(lookupWorkflowDeclaration _, lookupCall(jobKey, outputStore) _)
```

Comment:
hmm I'm not sure if we lost functionality or not here.  It'd be nice if `lookupWorkflowDeclaration` was instead something like `traverseScopeHierarchy(root=jobKey.call)`, and this could walk up the hierarchy (`root.ancestry`) looking for the variable.  I don't remember if we supported this before, but [scatter blocks can have declarations too](https://github.com/broadinstitute/wdl4s/blob/develop/src/main/resources/grammar.hgr#L339).

`traverseScopeHierarchy` could also be smart enough to address your comment above.

Maybe not something for this PR...